### PR TITLE
feat: Allow arrays of arbitrary types in the program ABI

### LINKED ISSUE #1459: Prover.toml doesn't handle arrays of structs
### Aim

Arrays of structs should handled

### Expected Behavior

Given the program
```rs
struct Foo {
    bar: Field,
    baz: Field,
}

fn main(foos: [Foo; 3]) -> pub Field {
    foos[2].bar + foos[2].baz
}
```
And the Prover.toml
```
[[foos]]
bar = 0
baz = 0

[[foos]]
bar = 0
baz = 0

[[foos]]
bar = 0
baz = 1
```
nargo should be able to prove

### Bug

Instead you encounter:
```
Error: input file is badly formed, could not parse, TOML parse error at line 3, column 1
  |
3 | [[foos]]
  | ^^^^^^^^
data did not match any variant of untagged enum TomlTypes
```

### To Reproduce

1. prove the above program and input

### Installation Method

Compiled from source

### Nargo Version

nargo 0.6.0 (git version hash: b695faa9127e12ea8adb2b4d46e19fd72a56051d, is dirty: true)

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #1459 

## Summary\*

This PR sets out to implement support for arrays of arbitrary types.

This requires us to make `InputValue` more flexible to allow encoding/decoding and we then need to expand this outwards

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [x] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
@joss-aztec were you running into this issue (#1459) when testing this with the new SSA? This currently panics due to the old SSA not supporting arrays of structs.

**Comment by jfecher** [comment]
> @joss-aztec Is this expected to work with the new SSA? This currently panics due to the old SSA not supporting arrays of structs.

It is expected to work since the new ssa is supposed to support arrays of structs and nested arrays, but we have been unable to test it to verify it works because these are not allowed in the frontend.

File: crates/noirc_abi/src/input_parser/json.rs [hunk 1]
```diff
@@ -79,8 +75,18 @@ impl From<InputValue> for JsonTypes {
...
+                    .iter()
+                    .map(|i| match i {
+                        InputValue::Field(field) => {
+                            JsonTypes::String(format!("0x{}", field.to_hex()))
+                        }
+                        _ => unreachable!(
+                            "Only arrays of simple field elements are allowable currently"
+                        ),
+                    })
+                    .collect();
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply crates/noirc_abi/src/input_parser/json.rs:88]: I'm currently locking this down for the time being to maintain the current behaviour while using the new `InputValue` and `TomlTypes` variants. Once I'm happy with these then I can loosen up the allowable types.

**Review by jfecher — APPROVED** [review]
Love it

**Pull request merged**